From: =?UTF-8?q?Leonard=20G=C3=B6hrs?= <l.goehrs@pengutronix.de>
Date: Mon, 11 Oct 2021 16:15:03 +0200
Subject: [PATCH] ARM: dts: stm32: add the ethernet port LEDs present on the
 LXATAC
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This commit adds previously un-exposed LEDs.
These LEDs are visible from the outside of the device, located on the device's
Ethernet Jacks and labeled "System Status" on the Enclosure.
The device enclosure has two sides indicated by the enclosure print,
the "DUT" side and the "LAB" side. The LEDs are named according to the
side of the device they are on.
These LEDs should show some activity when the device is plugged in as it is
otherwise not obvious if the device is powered.

led-2 is a normal active-low LED.

led-3 and led-4 illuminate the same light guide and are electrically connected
as follows:

 GPIOA14 --[1kΩ]--+-----------+
                  |           |
                  V (Green)   A (Orange)
                  |           |
 GPIOD15 ---------+-----------+

This means the LEDs will be on according to the following truth table:

 | GPIOA14 | GPIO15 | led-3 | led-4 |
 +---------+--------+-------+-------+
 | low     | low    | off   | off   |
 | high    | low    | on    | off   |
 | low     | high   | off   | on    |
 | high    | high   | off   | off   |

Ideally this would be handled by a multicolor LED driver that allows choosing
the LED color and maybe even allows setting the brightness via PWM.

For now this will do the right thing unless you try turning both LEDs on at the
same time (which can not work).

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>
---
 arch/arm/boot/dts/stm32mp157c-lxa-tac.dts | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts b/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
index 5171bb1e5037..ccf949eb80de 100644
--- a/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
+++ b/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
@@ -44,6 +44,31 @@ led-1 {
 			label = "tac:green:user2";
 			gpios = <&gpiog 7 GPIO_ACTIVE_HIGH>;
 		};
+
+		led-2 {
+			label = "tac:green:statusdut";
+			gpios = <&gpioa 13 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "default-on";
+		};
+
+		/* led-3 and led-4 are internally connected antiparallel to one
+		 * another inside the ethernet jack like this:
+		 * GPIOA14 ---+---|led-3|>--+--- GPIOD15
+		 *            +--<|led-4|---+
+		 * E.g. only one of the LEDs can be illuminated at a time while
+		 * the other output must be driven low.
+		 * This should likely be implemented using a multi color LED
+		 * driver for antiparallel LEDs. */
+		led-3 {
+			label = "tac:green:statuslab";
+			gpios = <&gpioa 14 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "default-on";
+		};
+
+		led-4 {
+			label = "tac:orange:statuslab";
+			gpios = <&gpiod 15 GPIO_ACTIVE_HIGH>;
+		};
 	};
 
 	gpio-keys {
